src/librustc_traits/lowering.rs

```diff
@@ -172,11 +172,19 @@ fn program_clauses_for_trait<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefI
     // ```
 
     // `FromEnv(WC) :- FromEnv(Self: Trait<P1..Pn>)`, for each where clause WC
-    // FIXME: Remove the [1..] slice; this is a hack because the query
+    // FIXME: Remove the filter; this is a hack because the query
     // predicates_of currently includes the trait itself (`Self: Trait<P1..Pn>`).
     let where_clauses = &tcx.predicates_of(def_id).predicates;
     let implied_bound_clauses =
-        where_clauses[1..].into_iter()
+        where_clauses.into_iter()
+        .filter(|wc| {
+            if let ty::Predicate::Trait(pred) = wc {
+                if pred.skip_binder().def_id() == def_id {
+                    return false;
+                }
+            }
+            true
+        })
```

Comment:
It seems like this would also skip a predicate of the form:
```rust
trait A where <Self as A>::Item: A {
    type Item;
}
```
or equivalently:
```rust
trait A {
    type Item: A;
}
```
since you are only relying on the def id. Actually maybe it would be better to just wait for your PR which removes the `Self: Trait` predicate...

Reply:
Ah, good catch. I'll try and make progress on that PR today. A patch that might possibly still work here is to filter out only the _first_ predicate that matches this condition, but it still feels quite brittle!